### DISCUSSION TITLE
v0.6.1 Pre-release thread

### DISCUSSION THREAD

**Discussion by masenf** [reflex-dev/reflex#4023] [Release]
Please report any problems you encounter in this discussion thread or as separate issues on reflex-dev/reflex.

```console
pip install 'reflex~=0.6.1a'
```

# Schedule

2024-09-28 09:40 PDT: 0.6.1a1 pre-release published for testing
2024-09-30 17:30 PDT: Public release of 0.6.1

# Release Notes

## Regression Fixes

* re add removed method with better behaviour and tests by @Lendemor in https://github.com/reflex-dev/reflex/pull/3986
* hash the state file name by @adhami3310 in https://github.com/reflex-dev/reflex/pull/4000
* EventFnArgMismatch fix to support defaults args by @LeoGrosjean in https://github.com/reflex-dev/reflex/pull/4004
* serialize default value for disk state manager by @adhami3310 in https://github.com/reflex-dev/reflex/pull/4008
* remove format_state and override behavior for bare by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3979
* Handle bool cast for optional NumberVar by @masenf in https://github.com/reflex-dev/reflex/pull/4010
* use literal var instead of serialize for toast props by @adhami3310 in #4027

### Experimental Dynamic Components

In this release, state vars can be of `rx.Component` type and `@rx.var` computed vars can return `rx.Component`, allowing dynamic component structures to be created on the backend at runtime. This enables the full use of python expressions in generated components, as opposed to `rx.cond` and `rx.foreach`.

There are performance implications when using dynamic components so prefer static components where possible.

* Component as Var type by @masenf in https://github.com/reflex-dev/reflex/pull/3732
* add basic integration test for dynamic components by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3990
* use bundled radix ui for dynamic components by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3993
* implement _evaluate in state by @adhami3310 in https://github.com/reflex-dev/reflex/pull/4018

### Run Backend with Granian server

install `granian` package and set `REFLEX_USE_GRANIAN=1` to run with the new, experimental granian backend (instead of the default uvicorn).

* can run with granian by setting REFLEX_USE_GRANIAN var by @Lendemor in https://github.com/reflex-dev/reflex/pull/3919
* change loglevel and fix granian on linux by @adhami3310 in https://github.com/reflex-dev/reflex/pull/4012

### `USE_SYSTEM_NODE` and `USE_SYSTEM_BUN`

If you prefer to run your own versions of `node` and `bun`, ensure they are on the `PATH` and set these environment variables to `1`.

* add env var to enable using system node and bun by @Lendemor in https://github.com/reflex-dev/reflex/pull/4006

### Bump nextjs to 14.2.13

* bump nextjs version by @Lendemor in https://github.com/reflex-dev/reflex/pull/3992

### Miscellaneous

* [ENG-3717] [flexgen] Initialize app from refactored code by @masenf in https://github.com/reflex-dev/reflex/pull/3918
* Added fill color for progress by @wassafshahzad in https://github.com/reflex-dev/reflex/pull/3926
* Use tailwind typography plugin by default by @minimav in https://github.com/reflex-dev/reflex/pull/3593
  * disable prose by default for rx.html by @adhami3310 in https://github.com/reflex-dev/reflex/pull/4001
  * use `class_name="prose"` to enable default formatting for `rx.html`
* allow link as metatags by @Lendemor in https://github.com/reflex-dev/reflex/pull/3980
* allow classname to be state vars by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3991
* default on_submit in form set to prevent_default by @Lendemor in https://github.com/reflex-dev/reflex/pull/4005

## Bug Fixes

* [ENG-3824]Make AI template use light mode by @ElijahAhianyo in https://github.com/reflex-dev/reflex/pull/3963
* don't camel case keys of dicts in style by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3982
* add missing message when running in backend_only by @Lendemor in https://github.com/reflex-dev/reflex/pull/4002
* [ENG-3849] Track backend-only vars that are declared without a default value by @masenf in https://github.com/reflex-dev/reflex/pull/4016
* improve lifespan typecheck and debug by @Lendemor in https://github.com/reflex-dev/reflex/pull/4014

## Documentation

* fix: Adding in-line comments for the segmented control: value and on_change by @elviskahoro in https://github.com/reflex-dev/reflex/pull/3933
* fix: Adding code comments for segmented control type by @elviskahoro in https://github.com/reflex-dev/reflex/pull/3935

## Other Changes

* add some unit tests for coverage by @Lendemor in https://github.com/reflex-dev/reflex/pull/3947
* always print passed_type by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3950
* remove unused badge by @Lendemor in https://github.com/reflex-dev/reflex/pull/3955
* add some more tests by @Lendemor in https://github.com/reflex-dev/reflex/pull/3965
* use svg elements instead of raw html for logo by @Lendemor in https://github.com/reflex-dev/reflex/pull/3978
* bump to 0.6.1 for further dev by @masenf in https://github.com/reflex-dev/reflex/pull/3995
* reorganize all tests in a single top folder by @Lendemor in https://github.com/reflex-dev/reflex/pull/3981
* Update markdown component map to use new rx.code_block.theme enum by @masenf in https://github.com/reflex-dev/reflex/pull/3996
* remove all runtime asserts by @Lendemor in https://github.com/reflex-dev/reflex/pull/4019

## New Contributors
* @minimav made their first contribution in https://github.com/reflex-dev/reflex/pull/3593
* @LeoGrosjean made their first contribution in https://github.com/reflex-dev/reflex/pull/4004

**Full Changelog**: https://github.com/reflex-dev/reflex/compare/v0.6.0...reflex-0.6.1

**Comment by bertbarabas** [comment]
My app mostly works on 0.6.1a but I did get the following exception printed in the terminal while uploading a 500MB file. It did not seem to cause my app to fail, just the exception was printed:

[ERROR] Application callable raised an exception
Traceback (most recent call last):
  File ".venv/lib/python3.12/site-packages/granian/_futures.py", line 4, in future_watcher
    await inner(watcher.scope, watcher.proto)
  File ".venv/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File ".venv/lib/python3.12/site-packages/starlette/middleware/cors.py", line 93, in __call__
    await self.simple_response(scope, receive, send, request_headers=headers)
  File ".venv/lib/python3.12/site-packages/starlette/middleware/cors.py", line 144, in simple_response
    await self.app(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 51, in wrapped_app
    await app(scope, receive, sender)
  File ".venv/lib/python3.12/site-packages/starlette/routing.py", line 715, in __call__
    await self.middleware_stack(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/routing.py", line 735, in app
    await route.handle(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/routing.py", line 288, in handle
    await self.app(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/routing.py", line 76, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 51, in wrapped_app
    await app(scope, receive, sender)
  File ".venv/lib/python3.12/site-packages/starlette/routing.py", line 74, in app
    await response(scope, receive, send)
  File ".venv/lib/python3.12/site-packages/starlette/responses.py", line 250, in __call__
    async with anyio.create_task_group() as task_group:
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 740, in __aexit__
    raise exc_val
  File ".venv/lib/python3.12/site-packages/starlette/responses.py", line 257, in __call__
    await wrap(partial(self.listen_for_disconnect, receive))
  File ".venv/lib/python3.12/site-packages/starlette/responses.py", line 253, in wrap
    await func()
  File ".venv/lib/python3.12/site-packages/starlette/responses.py", line 230, in listen_for_disconnect
    message = await receive()
              ^^^^^^^^^^^^^^^
CancelledError: Future cancelled.

**Reply by bertbarabas** [reply]
The exception only happened (so far) if I launch with:
REFLEX_USE_GRANIAN=1 reflex run

It did not happen with 
REFLEX_USE_GRANIAN=0 reflex run

**Reply by masenf** [reply]
filed as #4032

**Comment by masenf** [comment]
v0.6.1 is released
